src/diff.rs

```diff
@@ -61,11 +61,75 @@ pub struct ModifiedFile {
 
 #[derive(Debug, PartialEq, Eq)]
 pub struct FileDiff {
-    // TODO
+    pub hunks: Vec<Hunk>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Hunk {
+    pub header: Line,
+    pub lines: Vec<LineDiff>,
+}
+
+/// The content of a single line.
+pub type Line = Vec<u8>;
+
+/// Single line delta. Two of these are need to represented a modified line: one
+/// addition and one deletion. Context is also represented with this type.
+#[derive(Debug, PartialEq, Eq)]
+pub struct LineDiff {
+    /// Line number in old file or `None` for added line.
+    pub old_line_num: Option<u32>,
+    /// Line number in new file or `None` for deleted line.
+    pub new_line_num: Option<u32>,
+    /// Line content.
+    pub line: Line,
+    /// Sigil showing origin of this `LineDiff`. Eg. `+`, `-`, etc.
```

Comment:
What if we made an enum here instead:
```rust
enum LineDiff {
  Addition { line_num: u32, line: Line }, // this is the new line
  Removal { line_num: u32, line: Line }, // this was the old line
  Modification { line_num: u32, old: Line, new: Line },
}
```

This way we're not matching on a `char` :grin: (well apart from when we parse)

Reply:
Oh and `Context` too.

Reply:
So I thought about this, but we can actually match on the old/new line:

`(Some(old), None)` is a deletion
`(None, Some(new))` is an addition
`(Some(old), Some(new))` is context

Putting the whole thing in an enum makes it a bit annoying to retrieve the lines. Also keep in mind that there is no `Modification`, because those would be represented as *two* `LineDiff`s.

Reply:
So we'd end up with:

```rust
enum LineDiff {
  Addition { line_num: u32, line: Line }, // this is the new line
  Removal { line_num: u32, line: Line }, // this was the old line
  Context { line_num: u32, line: Line },
}
```

which has the benefit of being able to do:
```
match line_diff {
  Removal { line_num } | Context { line_num } | Addition { line_num } => ...
}
```

But it could just as well be represented as:
```
struct LineDiff {
  line: Line,
  kind: LineDiffKind
}
enum LineDiffKind {
  Addition, Deletion, Context
}
```
:thinking:

Reply:
I feel like remembering the order of `Some` and `None` in the tuple is a bit of a burden on the user :sweat_smile: And good point about `Modification` :)

I usually opt for the former because it's in canonical form -- sum of products. As opposed to the latter which is product of sums. But I don't think matters much.

My main goal here is to show the intent of the data through the types, so as long as we're describing the cases in a closed way I'll be happy and so will our consumers :)